### PR TITLE
Multibranch conditionals

### PR DESCRIPTION
**Pull request by seanchen1991**

### CODE REVIEW

File: src/ast/visitor.rs [hunk 1]
```diff
@@ -76,10 +76,12 @@ pub trait NodeVisitor<'a> {
 
     fn super_cond(
         &mut self,
-        if_blk: &Located<Block<'a>>,
-        do_blk: &Located<Block<'a>>,
+        branch: &Branch<'a>,
         el_blk: &Located<Block<'a>>,
     ) {
+        let if_blk = &branch.cond;
```
Comment (pvdrz) [inline src/ast/visitor.rs]: Could you add a `visit_branch` and `super_branch` methods to the trait? I think that'd be easier to maintain.

File: src/ast/mod.rs [hunk 2]
```diff
@@ -124,6 +124,12 @@ impl<'a> Display for Primitive {
     }
 }
 
+#[derive(Debug, Eq, PartialEq)]
+pub struct Branch<'a> {
```
Comment (pvdrz) [inline src/ast/mod.rs:129]: I know we don't have a lot of docs but could you add an small description of this structure?

File: src/ast/visitor.rs [hunk 3]
```diff
@@ -76,15 +76,32 @@ pub trait NodeVisitor<'a> {
...
     ) {
-        self.visit_block(&if_blk.content);
-        self.visit_block(&do_blk.content);
+        self.visit_branch(if_branch);
+
+        for branch in branches {
+            self.visit_branch(branch);
+        }
+
+        // let if_blk = &if_branch.cond;
```
Comment (pvdrz) [inline src/ast/visitor.rs]: Clean these when you're done 

(I usually forget to do that kind of stuff)

File: src/parser/node/cond.rs [hunk 4]
```diff
@@ -36,13 +37,14 @@ pub fn cond(input: Span) -> IResult<Located<Node>> {
             position,
             if_block,
             do_block,
+            many0(pair(elif_block, do_block)),
```
Comment (christianpoveda) [inline src/parser/node/cond.rs:40]: remove this line for now

File: src/parser/node/cond.rs [hunk 5]
```diff
@@ -36,13 +37,14 @@ pub fn cond(input: Span) -> IResult<Located<Node>> {
...
             // FIXME: fix optional else block
             else_block,
             preceded(keyword("end"), position),
         )),
-        move |(sp1, if_block, do_block, else_block, sp2)| {
+        move |(sp1, if_block, do_block, elif_block, else_block, sp2)| {
```
Comment (christianpoveda) [inline src/parser/node/cond.rs:45]: same here, remove the new `elif_block`

File: tests/parse/mod.rs [hunk 6]
```diff
@@ -281,24 +281,69 @@ fn cond() -> LangResult<'static, ()> {
...
-        .loc(),
+        ).loc(),
     ];
 
     assert_eq!(expected[0], result[0], "simple blocks");
     assert_eq!(expected[1], result[1], "long blocks");
     Ok(())
 }
 
+// #[test]
```
Comment (pvdrz) [inline tests/parse/mod.rs]: Also remove the commented test (but plz save it somewhere else, there is no way anyone is going to rewrite that huge AST :P)

File: tests/parse/elif.pj [hunk 7]
```diff
@@ -0,0 +1,23 @@
+if x do y elif a do b else z end
```
Comment (pvdrz) [inline tests/parse/elif.pj:1]: same for this file. But save it somewhere else

**Review by pvdrz — CHANGES_REQUESTED** [review]

File: src/parser/node/cond.rs [hunk 8]
```diff
@@ -4,20 +4,21 @@
 //! rule
 //!
 //! ```abnf
-//! cond = "if" block1 "do" block1 ("else" block1)? "end"
+//! cond = "if" block1 "do" block1 ("elif" block1 "do" block1) ("else" block1)? "end"
```
Comment (pvdrz) [inline src/parser/node/cond.rs]: change `("elif" block1 "do" block1)` to  `("elif" block1 "do" block1)*` to indicate that there could be zero or more of those.

Also change the `("else" block1)?` by just `"else" block1`. That used to be the grammar when we had optional `else`s but we no longer do that and I forgot to update the docs :/

File: src/parser/node/cond.rs [hunk 9]
```diff
@@ -34,47 +35,54 @@ pub fn cond(input: Span) -> IResult<Located<Node>> {
...
-/// Parses the `do` block of a [`Node::Cond`].
-///
-/// There must be at least one space or line break between the `do` and the first node in the
-/// block. There can be spaces or line breaks at the end of the block.
-///
-/// The location of the returned block ignores the `do` and spaces surrounding the block.
-fn do_block(input: Span) -> IResult<Located<Block>> {
-    delimited(keyword_space("do"), block1, multispace0)(input)
-}
+// pub fn cond(input: Span) -> IResult<Located<Node>> {
```
Comment (pvdrz) [inline src/parser/node/cond.rs]: remove this dead code

File: src/parser/node/cond.rs [hunk 10]
```diff
@@ -4,20 +4,21 @@
 //! rule
 //!
 //! ```abnf
-//! cond = "if" block1 "do" block1 ("else" block1)? "end"
+//! cond = "if" block1 "do" block1 ("elif" block1 "do" block1) ("else" block1)? "end"
 //! ```
 //!
-//! Thus, `else` blocks are optional and are represented as empty [`Block`]s inside the
+//! Thus, `elif` and `else` blocks are optional and are represented as empty [`Block`]s inside the
```
Comment (pvdrz) [inline src/parser/node/cond.rs]: plz remove the "and `else`". That's a leftover of some changes I did and forgot to update the docs :/

**Review by pvdrz — CHANGES_REQUESTED** [review]

**Review by pvdrz — APPROVED** [review]

**Pull request merged**